### PR TITLE
PPS lhcInfo split: LHCInfoPerLS and LHCInfoPerFill records

### PR DESCRIPTION
**Pull request by JanChyczynski**
#### PR description:

Definition of the new records, `LHCInfoPerLS` and `LHCInfoPerFill`, to be used instead of LHCInfo (to save space in the database by migrating members with different granularity to separate classes). 

`LHCInfoPerLS` contains, among others, `betaStarX/Y` and `crossingAngleX/Y` (these fields were separated for X and Y) used mainly in PPS proton reconstruction . `LHCInfoPerFill` is a refactored clone of `LHCInfo` with `lumiSection`, `betaStar` and `crossingAngle` removed.
    
`LHCInfoVectorizedFields` is a base class for `LHCInfoPerLS` and `LHCInfoPerFill` handling fields, with values stored in vectors for expandability purposes.

#### PR validation:

Validation by scripts creating objects of the new classes, writing to a local sqlite database, retrieving and displaying them on stdout by the print method:

```
cd CondTools/RunInfo/test
cmsRun LHCInfoPerLSWriter_cfg.py 
cmsRun LHCInfoPerFillWriter_cfg.py 
cmsRun LHCInfoPerLSAnalyzer_cfg.py 
cmsRun LHCInfoPerFillAnalyzer_cfg.py 
```

Additionally serialization unit tests were added to `testSerializationRunInfo.cpp`.

The new clases were also used in further development (Populator of Conditions, update of proton reconstruction; to be submitted in next PRs) which confirmed no problems with the definitions of the records.

### CODE REVIEW

**Comment by ChrisMisan** [comment]
type ctpps

**Comment by ChrisMisan** [comment]
@JanChyczynski you've deleted the part of the pr template, please specify the intentions for backports.

> Validation by scripts creating objects of the new classes, writing to a local sqlite database, retrieving and displaying them on stdout by the print method

Can this be made into unit test?

**Comment by ChrisMisan** [comment]
please test

File: CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ Usage:
+    <usage>
+
+*/
+//
+// Author:
+// Created:     Tue Feb 15 2022
+// $Id$
+//
+
```
Comment (grzanka) [inline CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h:21]:
```suggestion
```

do we need this ?

File: CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//
+// Package:     DataRecord
+// Class  :     LHCInfoPerLSRcd
+//
+/**\class LHCInfoPerLSRcd LHCInfoPerLSRcd.h CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h
+
+ Description: [one line class summary]
+
+ Usage:
+    <usage>
```
Comment (grzanka) [inline CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h:13]: Can you add few words here ?

File: CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+//
+// Package:     CondFormats/DataRecord
+// Class  :     LHCInfoPerFillRcd
+//
+/**\class LHCInfoPerFillRcd LHCInfoPerFillRcd.h CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h
+
+ Description: [one line class summary]
+
+ Usage:
+    <usage>
```
Comment (grzanka) [inline CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h:13]: Can you add few words here ?

File: CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+*/
+//
+// Author:
+// Created:     Tue Feb 15 2022
+// $Id$
+//
```
Comment (grzanka) [inline CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h:20]:
```suggestion
```

File: CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+*/
+//
+// Author:      Jan Krzysztof Chyczynski
+// Created:     Wed, 13 Jul 2022 16:36:24 GMT
+//
```
Comment (grzanka) [inline CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h:19]:
```suggestion
```

authorship stays in git as well

**Comment by JanChyczynski** [comment]
> 



> @JanChyczynski you've deleted the part of the pr template, please specify the intentions for backports.
> 
> > Validation by scripts creating objects of the new classes, writing to a local sqlite database, retrieving and displaying them on stdout by the print method
> 
> Can this be made into unit test?

For the records in CondFormats/RunInfo there are only serialization unit tests in `testSerializationRunInfo.cpp` which I extended for the new records. Is that enough?

**Comment by ChrisMisan** [comment]
@JanChyczynski I was talking abot these:

> cmsRun LHCInfoPerLSWriter_cfg.py 
cmsRun LHCInfoPerFillWriter_cfg.py 
cmsRun LHCInfoPerLSAnalyzer_cfg.py 
cmsRun LHCInfoPerFillAnalyzer_cfg.py

**Comment by ChrisMisan** [comment]
I am assuming this does not need to be backported?

**Comment by JanChyczynski** [comment]
> I am assuming this does not need to be backported?

Probably, not. I've asked Giacomo just to be sure and I'll update the description when I get the answer.

**Comment by vavati** [comment]
We need this PR merged in order to submit the next steps. If there are no other issues please sign it. As for the backport this will be argument of discussion with ALCA.

**Comment by tvami** [comment]
> @JanChyczynski I was talking abot these:
> 
> 
> 
> > cmsRun LHCInfoPerLSWriter_cfg.py 
> 
> cmsRun LHCInfoPerFillWriter_cfg.py 
> 
> cmsRun LHCInfoPerLSAnalyzer_cfg.py 
> 
> cmsRun LHCInfoPerFillAnalyzer_cfg.py 

Let's make this happen and then we can sign it

File: CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
Comment (perrotta) [inline CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc:5]: Not needed?

Comment (perrotta) [inline CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc:5]: Not needed?

File: CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#include "FWCore/Framework/interface/ESWatcher.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+
+#include "CondCore/DBOutputService/interface/PoolDBOutputService.h"
```
Comment (perrotta) [inline CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc:12]: Not needed?

File: CondTools/RunInfo/plugins/LHCInfoPerFillWriter.cc [hunk 8]
```diff
@@ -0,0 +1,63 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
Comment (perrotta) [inline CondTools/RunInfo/plugins/LHCInfoPerFillWriter.cc:5]: Not needed?

File: CondTools/RunInfo/plugins/LHCInfoPerLSAnalyzer.cc [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
Comment (perrotta) [inline CondTools/RunInfo/plugins/LHCInfoPerLSAnalyzer.cc:5]: Not needed?

File: CondTools/RunInfo/plugins/LHCInfoPerLSWriter.cc [hunk 10]
```diff
@@ -0,0 +1,47 @@
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/one/EDAnalyzer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
Comment (perrotta) [inline CondTools/RunInfo/plugins/LHCInfoPerLSWriter.cc:5]: Not needed?

**Review by perrotta — COMMENTED** [review]
While answering the other review requests you can profit and clean up a bit the included headers

**Comment by grzanka** [comment]
> > @JanChyczynski I was talking abot these:
> > > cmsRun LHCInfoPerLSWriter_cfg.py
> > 
> > 
> > cmsRun LHCInfoPerFillWriter_cfg.py
> > cmsRun LHCInfoPerLSAnalyzer_cfg.py
> > cmsRun LHCInfoPerFillAnalyzer_cfg.py
> 
> Let's make this happen and then we can sign it

@tvami what would be the best place to put such unit test code ?
Here https://github.com/cms-sw/cmssw/tree/CMSSW_12_6_X/CondFormats/RunInfo ?

**Comment by tvami** [comment]
> > > @JanChyczynski I was talking abot these:
> > > > cmsRun LHCInfoPerLSWriter_cfg.py
> > > 
> > > 
> > > cmsRun LHCInfoPerFillWriter_cfg.py
> > > cmsRun LHCInfoPerLSAnalyzer_cfg.py
> > > cmsRun LHCInfoPerFillAnalyzer_cfg.py
> > 
> > 
> > Let's make this happen and then we can sign it
> 
> @tvami what would be the best place to put such unit test code ? Here https://github.com/cms-sw/cmssw/tree/CMSSW_12_6_X/CondFormats/RunInfo ?

Yes, `CondTools/RunInfo/test/` sounds good

File: CondTools/RunInfo/plugins/LHCInfoPerFillWriter.cc (hunk 8 above)
Comment (JanChyczynski) [reply CondTools/RunInfo/plugins/LHCInfoPerFillWriter.cc:5]: It doesn't compile without it

File: CondTools/RunInfo/plugins/LHCInfoPerLSAnalyzer.cc (hunk 9 above)
Comment (JanChyczynski) [reply CondTools/RunInfo/plugins/LHCInfoPerLSAnalyzer.cc:5]: It doesn't compile without it

File: CondTools/RunInfo/plugins/LHCInfoPerLSWriter.cc (hunk 10 above)
Comment (JanChyczynski) [reply CondTools/RunInfo/plugins/LHCInfoPerLSWriter.cc:5]: It doesn't compile without it

File: CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc (hunk 6 above)
Comment (JanChyczynski) [reply CondTools/RunInfo/plugins/LHCInfoPerFillAnalyzer.cc:5]: It doesn't compile without it

File: CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h (hunk 3 above)
Comment (JanChyczynski) [reply CondFormats/DataRecord/interface/LHCInfoPerFillRcd.h]: Done

File: CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h (hunk 2 above)
Comment (JanChyczynski) [reply CondFormats/DataRecord/interface/LHCInfoPerLSRcd.h]: Done

**Comment by JanChyczynski** [comment]
> > @tvami what would be the best place to put such unit test code ? Here https://github.com/cms-sw/cmssw/tree/CMSSW_12_6_X/CondFormats/RunInfo ?
> 
> Yes, `CondTools/RunInfo/test/` sounds good

The tests are done.

**Comment by ChrisMisan** [comment]
please test

**Review by ChrisMisan — COMMENTED** [review]
I am not convinced that the writer and analyzer modules should go under /plugins. Usually writer/analyzer modules can be used to write or print any payload of the class. In this case modules are writing and printing a specific test payload with dumb values, so those modules can only be used to test this pr and lack any functionality outside this purpose. @tvami what is your opinion on this?

**Comment by tvami** [comment]
From a first look, those seem fine in the plugins. 

@JanChyczynski but you def need to do code checks, and then squash them into 1 commit. We can only launch tests after that is done

**Comment by ChrisMisan** [comment]
please test

**Comment by ChrisMisan** [comment]
please test

**Comment by ChrisMisan** [comment]
please test

**Comment by ChrisMisan** [comment]
+alca

- differences only in msgLogger
- unit test passes

**Comment by ggovi** [comment]
+1

**Comment by perrotta** [comment]
+1
- About the code non compiling if you remove the (non used) ESHandler header include, I think that it is due to a few analyzers that does not include that header even if ESHandle's are uded in them. At some point one should move the include where it is actually needed, and not rely on some other included file that (uselessly) countains it

**Pull request merged**